ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/SnapServer.java

```diff
@@ -14,66 +14,619 @@
+                  endKeyBytes = range.endKeyHash();
+                  isPartialRange =
+                      !(startKeyBytes.equals(Hash.ZERO) && endKeyBytes.equals(HASH_LAST));
+                }
+
+                ArrayDeque<NavigableMap<Bytes32, Bytes>> collectedStorages = new ArrayDeque<>();
+                List<Bytes> proofNodes = new ArrayList<>();
+                final var worldStateProof =
+                    new WorldStateProofProvider(new WorldStateStorageCoordinator(storage));
+
+                for (var forAccountHash : range.hashes()) {
+                  var accountStorages =
+                      storage.streamFlatStorages(
+                          Hash.wrap(forAccountHash), startKeyBytes, endKeyBytes, statefulPredicate);
+
+                  //// address partial range queries that return empty
+                  if (accountStorages.isEmpty() && isPartialRange) {
+                    // fetch next slot after range, if it exists
+                    LOGGER.debug(
+                        "found no slots in range, taking first value starting from {}",
+                        asLogHash(range.endKeyHash()));
+                    accountStorages =
+                        storage.streamFlatStorages(
+                            Hash.wrap(forAccountHash), range.endKeyHash(), UInt256.MAX_VALUE, 1L);
+                  }
+
+                  // don't send empty storage ranges
+                  if (!accountStorages.isEmpty()) {
+                    collectedStorages.add(accountStorages);
+                  }
+
+                  // if a partial storage range was requested, or we interrupted storage due to
+                  // request limits, send proofs:
+                  if (isPartialRange || !statefulPredicate.shouldGetMore()) {
+                    // send a proof for the left side range origin
+                    proofNodes.addAll(
+                        worldStateProof.getStorageProofRelatedNodes(
+                            getAccountStorageRoot(forAccountHash, storage),
+                            forAccountHash,
+                            Hash.wrap(startKeyBytes)));
+                    if (!accountStorages.isEmpty()) {
+                      // send a proof for the last key on the right
+                      proofNodes.addAll(
+                          worldStateProof.getStorageProofRelatedNodes(
+                              getAccountStorageRoot(forAccountHash, storage),
+                              forAccountHash,
+                              Hash.wrap(accountStorages.lastKey())));
+                    }
+                  }
+
+                  if (!statefulPredicate.shouldGetMore()) {
+                    break;
+                  }
+                }
+
+                var resp = StorageRangeMessage.create(collectedStorages, proofNodes);
+                LOGGER.debug(
+                    "returned in {} storage {} to {} range {} to {} with {} storages and {} proofs, resp size {} of max {}",
```

Comment:
nit: this log message is duplicated, could be refactored to a method

Reply:
I am going to skip this one for now since all of the return messages at least subtly different.